src/main/java/edu/kit/compiler/semantic/DetailedNameTypeAstVisitor.java

```diff
@@ -0,0 +1,505 @@
+        case Void:
+            return false;
+        default:
+            throw new IllegalArgumentException("unsupported data type");
+        }
+    }
+
+    private void semanticError(AstObject object, String message, Object... args) {
+        object.setHasError(true);
+
+        // TODO: collect semantic errors in list
+        throw new SemanticException(
+            String.format(message, args),
+            object
+        );
+    }
+
+    public DataType visit(ProgramNode programNode) {
+        for (ClassNode _class : programNode.getClasses()) {
+            _class.accept(this);
+        }
+
+        return VoidType;
+    }
+
+    public DataType visit(ClassNode classNode) {
+        currentClassNamespace = Optional.empty();
+        symboltable.enterScope();
+
+        for (ClassNodeField field : classNode.getFields()) {
+            if (!isValidDataType(field.getType())) {
+                if (field.getType().getType() == DataTypeClass.Void) {
+                    semanticError(field, "void type is not allowed for a field");
+                } else {
+                    semanticError(field, "unknown reference type %s", field.getType().getRepresentation(stringTable));
+                }
+            }
+
+            if (!field.isHasError()) {
+                symboltable.insert(field);
+            }
+        }
+
+        for (MethodNode methodNode : classNode.getStaticMethods()) {
+            methodNode.accept(this);
+        }
+
+        currentClassNamespace = Optional.of(namespaceMapper.getClassNamespace(classNode));
+
+        for (MethodNode methodNode : classNode.getDynamicMethods()) {
+            methodNode.accept(this);
+        }
+
+        symboltable.leaveScope();
+        currentClassNamespace = Optional.empty();
+        return VoidType;
+    }
+
+    private DataType visitMethodNode(MethodNode methodNode) {
+        symboltable.enterScope();
+
+        if (!isValidDataType(methodNode.getType()) && methodNode.getType().getType() != DataTypeClass.Void) {
+            semanticError(methodNode, "unknown reference type %s", methodNode.getType().getRepresentation(stringTable));
+        }
+
+        for (MethodNodeParameter parameter : methodNode.getParameters()) {
+            if (!isValidDataType(parameter.getType())) {
+                if (parameter.getType().getType() == DataTypeClass.Void) {
+                    semanticError(parameter, "void type is not allowed for a method parameter");
+                } else {
+                    semanticError(parameter, "unknown reference type %s", parameter.getType().getRepresentation(stringTable));
+                }
+            }
+
+            if (!parameter.isHasError()) {
+                symboltable.insert(parameter);
+            }
+        }
+
+        expectedReturnType = methodNode.getType();
+        methodNode.getStatementBlock().accept(this);
+        expectedReturnType = null;
```

Comment:
You could get rid of this statefulness by adding an inner class which extends your visitor and has an extra field for the expected return type. There is only a single method that relies on this information, so you would only have to move that one into the inner class.

Reply:
If we want to avoid state, I would rather suggest to not use the visitor for iterating over the methods at all and instead do this in a separate function, creating a new visitor for each method. This is also what I'm doing for the third pass.